File: frontend/src/api/workspaces.ts
/**
 * 工作区接口（对应后端 `/api/v1/workspaces`，v0.15）。
 *
 * 工作区是 **Agent 的项目**：一个用户指定的根目录 + 一组知识库。
 * 它把"这个 Agent 在哪儿干活、能查哪些资料"绑在一起——
 * 所以 `kb_ids` 不是可选装饰，新会话默认继承它（见后端 conversations.py）。
 */

import { request } from './client'

export interface Workspace {
  id: string
  name: string
  /** 用户指定的根目录（绝对路径）。Agent 的文件操作被约束在这里。 */
  root_path: string
  description: string
  /** 这个工作区绑定的知识库。新会话默认继承它们。 */
  kb_ids: string[]
  conversation_count: number
  created_at: string | null
  updated_at: string | null
}

export interface WorkspacePayload {
  name: string
  root_path: string
  description?: string
  kb_ids?: string[]
}

export function listWorkspaces(): Promise<{ items: Workspace[] }> {
  return request<{ items: Workspace[] }>('/workspaces')
}

/** 目录浏览里的一行：一个子目录，或一个"起点"。 */
export interface DirectoryEntry {
  name: string
  path: string
  /** 能不能直接拿它当工作区。**与建工作区时同一份判定**，所以灰掉的也建不出来。 */
  selectable: boolean
  /** 不能选的原因（原样显示）。 */
  reason: string
  /**
   * 能不能在**这一层里**新建目录（v0.41）。
   *
   * 只有专用区域（`WorkspaceBrowse.area`）里为真：容器里除了数据目录几乎处处只读，
   * 与其让用户逐个试，不如把"能建"标在还能建的那一处、把"不能建"的原因标在别的行上。
   */
  creatable: boolean
  /** 不能在这里新建目录的原因（原样显示，界面把它摆在"新建文件夹"旁边）。 */
  create_reason: string
  /** 能不能给它改名（判定与 `renameDirectory` 被拒时同一份）。 */
  renamable: boolean
  /** 不能改名的原因（原样显示）。 */
  rename_reason: string
}

export interface WorkspaceBrowse {
  path: string
  /** **当前这一层自己**（名字 + 能不能选 + 能不能建 + 原因）。服务端给，界面不自己判。 */
  current: DirectoryEntry
  /** 上一级；已经在最上层时为 null。 */
  parent: string | null
  entries: DirectoryEntry[]
  /** 起点（**专用区域** / 家目录 / 盘符 / 已有工作区的目录）。 */
  roots: DirectoryEntry[]
  note: string
  /**
   * 专用可写区域（`<data_dir>/workspaces`）：选择器的默认落脚点，
   * 也是**唯一**能新建/改名目录的地方（v0.41）。
   *
   * 由服务端给而不是前端拼：数据目录在哪只有服务端知道。
   */
  area: string
}

/**
 * 列服务器上的目录（选工作区根目录用，v0.35）。
 *
 * **服务端的活**：工作区根目录是**服务器上**的路径，而浏览器里的目录选择器给的是
 * 客户端本机的东西——指向的是另一台机器。所以"选择"只能是"服务端列给你看"。
 * 管理员专属（成员建工作区只需填路径）。
 *
 * 不给 `path` 时落在**专用区域**：那是唯一能新建目录的地方，也是打开选择器时
 * 最该看到的地方（家目录在容器里往往只读甚至不存在）。
 */
export function browseDirectories(path?: string): Promise<WorkspaceBrowse> {
  const query = path ? `?path=${encodeURIComponent(path)}` : ''
  return request<WorkspaceBrowse>(`/workspaces/browse${query}`)
}

/**
 * 在服务器上新建一个目录（只建一层，重名会被拒）。
 *
 * **只在专用区域里能建**：`parent` 不在区域里时服务端会拒（422），
 * 而界面早就用那一行的 `creatable` / `create_reason` 把按钮灰掉了——
 * 两边判定同一份，所以不会"按钮亮着、点了却报错"。
 */
export function createDirectory(parent: string, name: string): Promise<DirectoryEntry> {
  return request<DirectoryEntry>('/workspaces/dirs', {
    method: 'POST',
    body: JSON.stringify({ parent, name }),
  })
}

/** 给服务器上的目录改名（**只改名，不搬位置**；同样只在专用区域里）。 */
export function renameDirectory(path: string, name: string): Promise<DirectoryEntry> {
  return request<DirectoryEntry>('/workspaces/dirs', {
    method: 'PATCH',
    body: JSON.stringify({ path, name }),
  })
}

export function createWorkspace(payload: WorkspacePayload): Promise<Workspace> {
  return request<Workspace>('/workspaces', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function updateWorkspace(
  workspaceId: string,
  payload: Partial<WorkspacePayload>,
): Promise<Workspace> {
  return request<Workspace>(`/workspaces/${workspaceId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  })
}

/** 删工作区。**里面的会话不会被删**，它们退回"未归档"那一栏。 */
export function deleteWorkspace(workspaceId: string): Promise<void> {
  return request<void>(`/workspaces/${workspaceId}`, { method: 'DELETE' })
}

File: backend/app/api/v1/workspaces.py
"""工作区端点（v0.15，设计见 ``docs/设计/Agent-工作区与能力层设计-v0.1.md``）。

工作区是 **Agent 的项目**：一个用户指定的根目录 + 一组知识库。
它把"这个 Agent 在哪儿干活、能查哪些资料"这两件事绑在一起。

**归属口径与知识库 / 会话完全一致**（``api/auth.py`` 的 ``check_kb_scope`` 同一套）：
普通成员只看得到自己的工作区，越权与不存在都回 **404**——403 会暴露"这个 id 存在"。

`root_path` 是这一组端点里唯一的安全边界（Agent 的文件操作会落在那里），
它的校验在 ``services/workspace.py::validate_root_path``，这里只做协议层转发。
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.api.auth import require_admin, require_read, require_write
from app.api.v1.schemas import (
    DirectoryEntryOut,
    WorkspaceBrowseOut,
    WorkspaceCreateIn,
    WorkspaceListOut,
    WorkspaceOut,
    WorkspaceUpdateIn,
)
from app.core.services import Services, get_services
from app.services.api_key import Caller

router = APIRouter(prefix="/workspaces", tags=["workspaces"])


def _owner(caller: Caller) -> str | None:
    """归属过滤用：只有**普通成员**会话才有归属；管理员与 API Key 通道没有。

    与 ``api/v1/notes.py::_owner``、``api/v1/conversations.py::_caller_owner``
    同一口径。三处各写一份是因为它们分属三个模块，但判定必须一致——
    `test_visibility_api` 那组用例正是钉这件事。
    """
    if caller.user is not None and not caller.is_admin:
        return caller.user.id
    return None


def _out(record, conversation_count: int) -> WorkspaceOut:  # type: ignore[no-untyped-def]
    return WorkspaceOut(
        id=record.id,
        name=record.name,
        root_path=record.root_path,
        description=record.description,
        kb_ids=list(record.kb_ids),
        conversation_count=conversation_count,
        created_at=record.created_at,
        updated_at=record.updated_at,
    )


@router.get("", response_model=WorkspaceListOut, summary="工作区列表")
def list_workspaces(
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_read)],
) -> WorkspaceListOut:
    views = services.workspaces.list(user_id=_owner(caller))
    return WorkspaceListOut(items=[_out(view.record, view.conversation_count) for view in views])


@router.get(
    "/browse",
    response_model=WorkspaceBrowseOut,
    summary="浏览服务器上的目录（选工作区根目录用）",
)
def browse_directories(
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_admin)],
    path: str | None = Query(default=None, description="要看哪个目录；留空 = 从家目录开始"),
) -> WorkspaceBrowseOut:
    """**管理员专属**：它列的是**服务器上**的目录树。

    这条与"设置页只认管理员"同一档：目录名本身就是信息（谁的项目叫什么、
    备份放在哪、有哪些账号的家目录），而成员建工作区本来就只需要填一个路径。
    换句话说是**不给它扩权**——能浏览不改变"能不能当工作区"的判定，
    那条判定只有一份（``workspaces.root_path_problem``）。

    只列**目录**；数据目录会出现在列表里但标着不可选与原因（不藏起来：
    静默省略会让人以为"这里没有它"，而他找的可能正是它旁边那个）。
    """
    view = services.workspaces.browse(path)
    return WorkspaceBrowseOut(
        path=view.path,
        current=DirectoryEntryOut(**asdict(view.current)),
        parent=view.parent,
        # `dataclasses.asdict` 而不是 `vars`：`DirectoryEntry` 是 slots 数据类，没有 `__dict__`
        entries=[DirectoryEntryOut(**asdict(item)) for item in view.entries],
        roots=[DirectoryEntryOut(**asdict(item)) for item in view.roots],
        note=view.note,
    )


@router.post(
    "",
    response_model=WorkspaceOut,
    status_code=status.HTTP_201_CREATED,
    summary="新建工作区（指定根目录）",
)
def create_workspace(
    payload: WorkspaceCreateIn,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
) -> WorkspaceOut:
    """`root_path` 必须是**已存在的目录**，且不能指向数据目录或文件系统根
    （见 ``validate_root_path`` 的三道校验）。"""
    record = services.workspaces.create(
        name=payload.name,
        root_path=payload.root_path,
        description=payload.description,
        kb_ids=payload.kb_ids,
        user_id=_owner(caller),
    )
    return _out(record, 0)


@router.get("/{workspace_id}", response_model=WorkspaceOut, summary="工作区详情")
def get_workspace(
    workspace_id: str,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_read)],
) -> WorkspaceOut:
    record = services.workspaces.get(workspace_id, user_id=_owner(caller))
    return _out(record, services.workspaces.conversation_count(workspace_id))


@router.patch("/{workspace_id}", response_model=WorkspaceOut, summary="改工作区")
def update_workspace(
    workspace_id: str,
    payload: WorkspaceUpdateIn,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
) -> WorkspaceOut:
    """名字 / 根目录 / 描述 / 知识库都可选，只改传了的那些。

    **归属不可改**：把一个工作区转给别人，连带的是"里头会话的 Agent 行为"，
    那是另一个功能，不该顺手做掉。
    """
    record = services.workspaces.update(
        workspace_id,
        user_id=_owner(caller),
        name=payload.name,
        root_path=payload.root_path,
        description=payload.description,
        kb_ids=payload.kb_ids,
    )
    return _out(record, services.workspaces.conversation_count(workspace_id))


@router.delete(
    "/{workspace_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除工作区（里面的会话退回未归档）",
)
def delete_workspace(
    workspace_id: str,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
) -> None:
    """**不删会话**：它们变成未归档，在侧栏的"未归档会话"那一栏继续存在。

    这是刻意的：会话里有用户问过的内容，误删不可恢复；而"失去归属"是可恢复的
    （重新挂一个工作区就行）。
    """
    services.workspaces.delete(workspace_id, user_id=_owner(caller))
